astropy/modeling/astro_sherpa.py

```diff
@@ -0,0 +1,433 @@
+            else:
+                _data = Data1DInt("wrapped_data", xlo=x - xerr, xhi=x + xerr, y=y, staterror=yerr)
+    else:
+        if xerr is None and yerr is None:
+            if zerr is None:
+                _data = Data2D("wrapped_data", x0=x, x1=y, y=z)
+            else:
+                _data = Data2D("wrapped_data", x0=x, x1=y, y=z, staterror=zerr)
+        elif xerr is not None and yerr is not None:
+            if zerr is None:
+                _data = Data2DInt("wrapped_data", x0lo=x - xerr, x0hi=x + xerr, x1lo=y - yerr, x1hi=y + yerr, y=z)
+            else:
+                _data = Data2DInt("wrapped_data", x0lo=x - xerr, x0hi=x + xerr, x1lo=y - yerr, x1hi=y + yerr, y=z, staterror=zerr)
+        else:
+            raise Exception("Set xerr and yerr, or set neither!")
+
+    return _data
+
+
+def make_datasets(n_dim, x, y, z=None, xerr=None, yerr=None, zerr=None):
+    '''
+    Parameters
+        ----------
+        n_dim: int
+            Used to veirfy required number of dimentions.
+        x : array (or list of arrays)
+            input coordinates
+        y : array (or list of arrays)
+            input coordinates
+        z : array (or list of arrays) (optional)
+            input coordinates
+        xerr : array (or list of arrays) (optional)
+            an array of errors in x
+        yerr : array (or list of arrays) (optional)
+            an array of errors in y
+        zerr : array (or list of arrays) (optional)
+            an array of errors in z
+
+    returns:
+        _data: a sherpa dataset
+    '''
+
+    x = np.array(x)
+    y = np.array(y)
+
+    if x.ndim == 2 or (x.dtype == np.object or y.dtype == np.object):
+        data = []
+        if z is None:
+            z = len(x) * [None]
+
+        if xerr is None:
+            xerr = len(x) * [None]
+
+        if yerr is None:
+            yerr = len(y) * [None]
+
+        if zerr is None:
+            zerr = len(z) * [None]
+
+        for nn, (xx, yy, zz, xxe, yye, zze) in enumerate(zip(x, y, z, xerr, yerr, zerr)):
+            data.append(_make_dataset(n_dim, x=xx, y=yy, z=zz, xerr=xxe, yerr=yye, zerr=zze, n=nn))
+        return DataSimulFit("wrapped_data", data)
+    else:
+        return _make_dataset(n_dim, x=x, y=y, z=z, xerr=xerr, yerr=yerr, zerr=zerr)
+
+
+def _make_simfit_model(models, tie_list, model_dict):
+    '''
+    makes a SimulFitModel for the models suplied it also links and parameters within tie_list
+    model_dict updated with astropy models and their converted conterparts this is used for
+    repopulating the model later
+    '''
+    for mod in models:
+        model_dict[mod] = _astropy_to_sherpa_model(mod)
+
+    if tie_list is not None:
+        for par1, par2 in tie_list:
+            getattr(model_dict[par1._model], par1.name).link = getattr(model_dict[par2._model], par2.name)
+    return SimulFitModel("wrapped_fit_model", model_dict.values())
```

Comment:
Ditto. Return an instance of a wrapper class so you don't need to call `isinstance` later on. Design the classes so you can just rely on their interface rather than on their identity.